Userland/Libraries/LibJS/Runtime/ArrayPrototype.cpp

```diff
@@ -836,21 +1094,43 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayPrototype::join)
         s_array_join_seen_objects.remove(this_object);
     };
 
+    // 2. Let len be ? LengthOfArrayLike(O).
     auto length = TRY(length_of_array_like(vm, this_object));
-    DeprecatedString separator = ",";
-    if (!vm.argument(0).is_undefined())
-        separator = TRY(vm.argument(0).to_deprecated_string(vm));
+
+    String separator_string;
+
+    // 3. If separator is undefined, let sep be ",".
+    if (separator.is_undefined())
+        separator_string = ","_short_string;
+    // 4. Else, let sep be ? ToString(separator).
+    else
+        separator_string = TRY(separator.to_string(vm));
+
+    // 5. Let R be the empty String.
     StringBuilder builder;
-    for (size_t i = 0; i < length; ++i) {
-        if (i > 0)
-            builder.append(separator);
-        auto value = TRY(this_object->get(i));
-        if (value.is_nullish())
-            continue;
-        auto string = TRY(value.to_deprecated_string(vm));
-        builder.append(string);
+
+    // 6. Let k be 0.
+    size_t k = 0;
+
+    // 7. Repeat, while k < len,
+    while (k < length) {
+        // a. If k > 0, set R to the string-concatenation of R and sep.
+        if (k > 0)
+            builder.append(separator_string);
+
+        // b. Let element be ? Get(O, ! ToString(𝔽(k))).
+        auto element = TRY(this_object->get(k));
+
+        // c. If element is either undefined or null, let next be the empty String; otherwise, let next be ? ToString(element).
+        // d. Set R to the string-concatenation of R and next.
+        if (!element.is_undefined() && !element.is_null())
+            builder.append(TRY(element.to_string(vm)));
+
+        // e. Set k to k + 1.
+        ++k;
     }
 
+    // 8. Return R.
     return PrimitiveString::create(vm, builder.to_deprecated_string());
```

Comment:
I tried to use `TRY` with `to_string` here, instead of the deprecated string.
But got some compiler error about errors not working with completions.
Is there a straightforward way to get that to work?

Reply:
`TRY_OR_THROW_OOM`

https://github.com/SerenityOS/serenity/blob/aa7ca80d7c1422c210b5a7109da46f232c255193/Userland/Libraries/LibJS/Runtime/Intl/LocalePrototype.cpp#L120

Reply:
Awesome, that's much cleaner :)